### PR TITLE
[CI] Enable CUDA SYCL CTS tests

### PR DESCRIPTION
**Pull request by yinyangsx**

### CODE REVIEW

File: devops/actions/khronos_cts_test/action.yml [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+        source /opt/runtimes/oneapi-tbb/env/vars.sh;
+      else
+        echo "no TBB vars in /opt/runtimes or /runtimes";
+      fi
+      # TODO remove workaround of FPGA emu bug
+      mkdir -p icd
+      echo /usr/lib/x86_64-linux-gnu/intel-opencl/libigdrcl.so > icd/gpu.icd
+      echo /runtimes/oclcpu/x64/libintelocl.so > icd/cpu.icd
+      echo /opt/runtimes/oclcpu/x64/libintelocl.so > icd/cpu2.icd
+      export OCL_ICD_VENDORS=$PWD/icd
```
Comment (bader) [inline devops/actions/khronos_cts_test/action.yml:105]: Why is this needed? I expect the container to be already correctly configured.

**Comment by bader** [comment]
Here is the summary of the SYCL-CTS execution on NVIDIA GPU.

Tests we can't build:
buffer
exceptions
multi_ptr
opencl_interop
reduction
group
kernel
math_builtin_api
nd_item
optional_kernel_features

Many test fail due to below error:
```
2022-07-26T03:36:28.2377097Z get_kernel_bundle_with_incompatible_kernels
2022-07-26T03:36:28.2377274Z -------------------------------------------------------------------------------
2022-07-26T03:36:28.2377387Z /__w/llvm/llvm/khronos_sycl_cts/tests/kernel_bundle/../common/../../util/proxy.h:35
2022-07-26T03:36:28.2377462Z ...............................................................................
2022-07-26T03:36:28.2377466Z 
2022-07-26T03:36:28.2377584Z /__w/llvm/llvm/khronos_sycl_cts/tests/kernel_bundle/../common/assertions.h:85: FAILED:
2022-07-26T03:36:28.2377646Z explicitly with message:
2022-07-26T03:36:28.2377762Z   No expected exception thrown for sycl::get_kernel_bundle(context, devices,
2022-07-26T03:36:28.2377833Z   kernel_ids) with bundle state input
2022-07-26T03:36:28.2377838Z 
2022-07-26T03:36:28.2377873Z 
2022-07-26T03:36:28.2377930Z PI CUDA ERROR:
2022-07-26T03:36:28.2377984Z                Value:           700
2022-07-26T03:36:28.2378054Z                Name:            CUDA_ERROR_ILLEGAL_ADDRESS
2022-07-26T03:36:28.2378152Z                Description:     an illegal memory access was encountered
2022-07-26T03:36:28.2378217Z                Function:        build_program
2022-07-26T03:36:28.2378326Z                Source Location: /__w/llvm/llvm/src/sycl/plugins/cuda/pi_cuda.cpp:703
2022-07-26T03:36:28.2378331Z 
2022-07-26T03:36:28.2378335Z 
2022-07-26T03:36:28.2378386Z PI CUDA ERROR:
2022-07-26T03:36:28.2378438Z                Value:           400
2022-07-26T03:36:28.2378512Z                Name:            CUDA_ERROR_INVALID_HANDLE
2022-07-26T03:36:28.2378590Z                Description:     invalid resource handle
2022-07-26T03:36:28.2378661Z                Function:        cuda_piProgramRelease
2022-07-26T03:36:28.2378770Z                Source Location: /__w/llvm/llvm/src/sycl/plugins/cuda/pi_cuda.cpp:3546
2022-07-26T03:36:28.2378775Z 
2022-07-26T03:36:28.2379013Z terminate called after throwing an instance of 'cl::sycl::runtime_error'
2022-07-26T03:36:28.2379294Z   what():  Native API failed. Native API returns: -5 (PI_ERROR_OUT_OF_RESOURCES) -5 (PI_ERROR_OUT_OF_RESOURCES)
```
@yinyangsx, I think something is wrong either with the HW or with the CUDA driver installation.
@AerialMantis, any other ideas?

@pvchupin, FYI.

File: devops/actions/khronos_cts_test/action.yml (hunk 1 above)
Comment (yinyangsx) [reply devops/actions/khronos_cts_test/action.yml]: This is what I have set up to mimic the environment of CUDA LLVM Test Suite, currently CUDA LLVM Test Suite is working well, so I think the environment should be correct. When the test is stable I will do this configuration when building the image.

Comment (bader) [reply devops/actions/khronos_cts_test/action.yml]: Ok. I'll do review when it's marked as ready for review then.

NOTE: Some of the tests you disabled should be fixed by https://github.com/KhronosGroup/SYCL-CTS/pull/367.
I suggest checking the status using latest SYCL-CTS source and re-enable disabled tests. Ideally we should use the filter file from the Khronos SYCL-CTS repository instead of creating a local duplicate.

**Review by bader — COMMENTED** [review]
I can't reproduce the issues from the latest CI test results on machine with NVIDIA GPU, but I don't use docker containers. So, I suspect it might be the reason of massive failures. Something wrong might with with configuring NVIDIA GPU device/driver inside a container.

**Comment by hdelan** [comment]
Hi @yinyangsx, am seeing if I can replicate some of these build/test run failures in a docker container. See preliminary results: 

| Test  | Docker image (GTX 1050) | Native DPCPP (A100) |
|-----|----------------|----------|
| buffer | Built successfully. No run failures but some tests skipped with warning [1] | Built & run successfully |
| group | Built successfully. Failed run with error [1] | Built and run successfully |
| kernel | Built successfully. Failed run with error [1] | Built and run successfully |
| math_builtin_api | Built successfully. All tests pass | Built successfully. One test fails with failure [1] |
| nd_item | Built and run successfully |  Built and run successfully |




Errors:
[1] `/tmp/dpcpp/sycl/source/detail/platform_impl.hpp:115: const cl::sycl::detail::plugin& cl::sycl::detail::platform_impl::getPlugin() const: Assertion !MHostPlatform && "Plugin is not available for Host."' failed.`


Warnings:
[1] `/SYCL-CTS/util/logger.cpp:41: warning: Device does not support double precision floating point operations`

Failures:
[1] 
```
warning:
     value: inf [7f800000], reference: 0 [0]
   
   /media/storage/hughdelaney/SYCL-CTS/util/logger.cpp:29: 
   warning:
     Expected accuracy in ULP: 16
   
   /media/storage/hughdelaney/SYCL-CTS/tests/math_builtin_api/math_builtin.h:149: FAILED:
   explicitly with message:
     tests case: 4000359. Correctness check failed.
```

**Comment by hdelan** [comment]
So I think there is most likely something wrong with the docker configuration.

**Comment by bader** [comment]
@hdelan, could you share the output of `sycl-ls` for "Native DPCPP (A100)" system, please?

**Comment by hdelan** [comment]
Native DPCPP:
```
$ sycl-ls
[ext_oneapi_cuda:gpu:0] NVIDIA CUDA BACKEND, NVIDIA A100-PCIE-40GB 0.0 [CUDA 11.6]
[ext_oneapi_cuda:gpu:1] NVIDIA CUDA BACKEND, NVIDIA A100-PCIE-40GB 0.0 [CUDA 11.6]
[host:host:0] SYCL host platform, SYCL host device 1.2 [1.2]
```

Docker:
```
root@9cb8282cce9a:/# sycl-ls
[ext_oneapi_cuda:gpu:0] NVIDIA CUDA BACKEND, NVIDIA GeForce GTX 1050 Ti 0.0 [CUDA 11.7]
[host:host:0] SYCL host platform, SYCL host device 1.2 [1.2]
```

**Comment by hdelan** [comment]
I had an issue initially where the docker image did not pick up the correct headers when compiling SYCL-CTS within the docker container, (which was not a problem for other compilations). I used a quick hack to fix this with 
```
RUN ln -s /$DPCPP_BUILD_DIR/include/sycl /include
```
If build failures are occurring due to missing headers this may be something to try.

Also the `OpenCL_INCLUDE_DIR=$DPCPP_INSTALL_DIR/include/sycl/CL` must be set when configuring the project

**Comment by bader** [comment]
I think we can avoid this error:
/tmp/dpcpp/sycl/source/detail/platform_impl.hpp:115: const cl::sycl::detail::plugin& cl::sycl::detail::platform_impl::getPlugin() const: Assertion !MHostPlatform && "Plugin is not available for Host."' failed.

If we configure OpenCL CPU runtime.

For fixing a warning, we need a GPU which supports doubles.

math_builtin_api test failure deserves investigation. Maybe we some FP controls allow unsafe FP computations.

**Comment by bader** [comment]
> Also the `OpenCL_INCLUDE_DIR=$DPCPP_INSTALL_DIR/include/sycl/CL` must be set when configuring the project

The script installs OpenCL headers: https://github.com/intel/llvm/pull/6439/files#diff-b1ff31309be0238339d92277f6849d504bf6059c74f073af63670edbdaca5e87R33

> I had an issue initially where the docker image did not pick up the correct headers when compiling SYCL-CTS within the docker container, (which was not a problem for other compilations).

I don't see any issues like this in the runs. There should be a lot of build issues if DPC++ headers are not found. The build is successful, but execution fails.

I don't see CUDA SDK installed in the docker image, so I assume it's already installed on the system, but it enabling GPU device in Docker image might require additional configuration. Did you expose additional devices or system directories to the docker image?

**Comment by hdelan** [comment]
> If we configure OpenCL CPU runtime.

This is true, output to be updated once tests finish running.

> For fixing a warning, we need a GPU which supports doubles.

The GTX1050 does support doubles, will compare with native DPCPP on GTX1050 to see if tests similarly skipped.

> math_builtin_api test failure deserves investigation. Maybe we some FP controls allow unsafe FP computations.

Will investigate this

**Comment by hdelan** [comment]
> I don't see CUDA SDK installed in the docker image, so I assume it's already installed on the system, but it enabling GPU device in Docker image might require additional configuration. Did you expose additional devices or system directories to the docker image?

I had assumed that https://github.com/intel/llvm/pull/6439/files#diff-e73ff2bdc5774902079806eea75725109964ea0442652ef7ac4499a1abbf9e3cR84 this line would pick up an image with a cuda installation. If the CUDA SDK is not included then it can be mounted using `-v`. However, I had assumed that the nvidia docker image contains necessary drivers as well as the CUDA SDK, so it is worthwhile making sure that `nvidia/cuda:*-devel-ubuntu22.04` or something similar is being used.

**Comment by bader** [comment]
> > I don't see CUDA SDK installed in the docker image, so I assume it's already installed on the system, but it enabling GPU device in Docker image might require additional configuration. Did you expose additional devices or system directories to the docker image?
> 
> I had assumed that https://github.com/intel/llvm/pull/6439/files#diff-e73ff2bdc5774902079806eea75725109964ea0442652ef7ac4499a1abbf9e3cR84 this line would pick up an image with a cuda installation. If the CUDA SDK is not included then it can be mounted using `-v`. However, I had assumed that the nvidia docker image contains necessary drivers as well as the CUDA SDK, so it is worthwhile making sure that `nvidia/cuda:*-devel-ubuntu22.04` or something similar is being used.

cuda_image is our own built container - https://github.com/intel/llvm/blob/sycl/.github/workflows/sycl_resolve_test_matrix.yml#L14-L17, which is based on NVIDIA's container - https://github.com/intel/llvm/blob/sycl/devops/containers/ubuntu2004_build.Dockerfile#L1.

What's confusing is that we use similar configuration for running llvm-test-suite on this machine and there are no such issues there.

**Comment by hdelan** [comment]
It is strange. I am assuming that `nvidia-container-runtime` is installed on the machine running the docker container, although I cannot see any mention of it in the CI?

Also: the precision problems on A100 are related to ptxas-11.6. They may have been fixed in SDK 11.7 but I don't have a machine I can check with. I have submitted a bug report to Nvidia.

**Comment by pvchupin** [comment]
@hdelan, nvidia-container-runtime is installed on the host machine that runs a container. It's part of the initial host setup procedure.

**Comment by hdelan** [comment]
Thanks @pvchupin . @yinyangsx are the failures still the same? What is the output of the build failures?

**Comment by yinyangsx** [comment]
> Thanks @pvchupin . @yinyangsx are the failures still the same? What is the output of the build failures?

Hi @hdelan, from the CI log, the run fails at building SYCL-CTS. Here is the error output,
2022-08-02T18:00:34.1375581Z FAILED: bin/test_math_builtin_api 
2022-08-02T18:00:34.1378581Z : && /__w/llvm/llvm/toolchain/bin/clang++ -DSYCL2020_CONFORMANT_APIS -DSYCL2020_DISABLE_DEPRECATION_WARNINGS -D__SYCL_DISABLE_PARALLEL_FOR_RANGE_ROUNDING__ -ffp-model=precise      -fsycl -sycl-std=2020 -fsycl-targets=nvptx64-nvidia-cuda tests/math_builtin_api/CMakeFiles/test_math_builtin_api_objects.dir/math_builtin_common_base.cpp.o tests/math_builtin_api/CMakeFiles/test_math_builtin_api_objects.dir/math_builtin_common_half.cpp.o tests/math_builtin_api/CMakeFiles/test_math_builtin_api_objects.dir/math_builtin_common_double.cpp.o tests/math_builtin_api/CMakeFiles/test_math_builtin_api_objects.dir/math_builtin_float_base.cpp.o tests/math_builtin_api/CMakeFiles/test_math_builtin_api_objects.dir/math_builtin_float_half.cpp.o tests/math_builtin_api/CMakeFiles/test_math_builtin_api_objects.dir/math_builtin_float_double.cpp.o tests/math_builtin_api/CMakeFiles/test_math_builtin_api_objects.dir/math_builtin_relational_base.cpp.o tests/math_builtin_api/CMakeFiles/test_math_builtin_api_objects.dir/math_builtin_relational_half.cpp.o tests/math_builtin_api/CMakeFiles/test_math_builtin_api_objects.dir/math_builtin_relational_double.cpp.o tests/math_builtin_api/CMakeFiles/test_math_builtin_api_objects.dir/math_builtin_geometric_base.cpp.o tests/math_builtin_api/CMakeFiles/test_math_builtin_api_objects.dir/math_builtin_geometric_double.cpp.o tests/math_builtin_api/CMakeFiles/test_math_builtin_api_objects.dir/math_builtin_integer.cpp.o tests/math_builtin_api/CMakeFiles/test_math_builtin_api_objects.dir/math_builtin_native.cpp.o tests/math_builtin_api/CMakeFiles/test_math_builtin_api_objects.dir/math_builtin_half.cpp.o tests/common/CMakeFiles/main_function_object.dir/main.cpp.o -o bin/test_math_builtin_api  util/libutil.a  oclmath/liboclmath.a  vendor/Catch2/src/libCatch2.a  /usr/lib/x86_64-linux-gnu/libOpenCL.so  -lpthread && :
2022-08-02T18:00:34.1380392Z llvm-foreach: Segmentation fault (core dumped)
2022-08-02T18:00:34.1380690Z clang-16: error: ptxas command failed with exit code 254 (use -v to see invocation)
2022-08-02T18:00:34.1381229Z clang version 16.0.0 (https://github.com/intel/llvm 6e1ff26a0385d57e9c05228ce23c2b9d785d2c64)
2022-08-02T18:00:34.1381481Z Target: x86_64-unknown-linux-gnu
2022-08-02T18:00:34.1381629Z Thread model: posix

**Comment by yinyangsx** [comment]
> Thanks @pvchupin . @yinyangsx are the failures still the same? What is the output of the build failures?

Hi @hdelan, do you want me to restart the CI tests? Still need me to make some changes and restart the test?

**Comment by hdelan** [comment]
> 

Aha I think I might know the cause here. What CUDA version is being used on the CI machine? There was a problem with CUDA SDK 11.4 (and maybe 11.3 as well) where the same math tests were making ptxas segfault. See https://github.com/intel/llvm/issues/4436 . Updating the CUDA version to a more recent version should fix this.

**Comment by hdelan** [comment]
In ubuntu2004_build.Dockerfile the line:
```
FROM nvidia/cuda:11.4.2-devel-ubuntu20.04
```
Should be changed to 11.6.** or something like that

**Comment by hdelan** [comment]
@yinyangsx Once these changes are made it would be great if you could restart the CI

**Comment by bader** [comment]
> @yinyangsx Once these changes are made it would be great if you could restart the CI

I've created https://github.com/intel/llvm/pull/6545 to update CUDA version.
@yinyangsx, please, resolve merge conflicts for this PR.

**Comment by bader** [comment]
https://github.com/intel/llvm/pull/6545 is merged. I expect next pre-commit to use CUDA 11.7.

**Comment by yinyangsx** [comment]
> > @yinyangsx Once these changes are made it would be great if you could restart the CI
> 
> I've created #6545 to update CUDA version. @yinyangsx, please, resolve merge conflicts for this PR.

Resolved merge conflicts and restarted CI tests.

**Comment by bader** [comment]
Thanks. Unfortunately CUDA machine is offline, so there will be delay with getting updated results.

**Comment by hdelan** [comment]
This PR fixes some issues with mem_advise tests https://github.com/intel/llvm/pull/6567

**Comment by bader** [comment]
@yinyangsx, I see that each week SYCL-CTS has new regressions. To test GHA scripts, I recommend disabling all the tests except one (e.g. test_header or test_id) and re-enable them back after automation scripts are committed. Otherwise, we won't be able to commit this soon. In addition to that building the whole test suite takes significant amount of time, which is not really required for testing GHA scripts.

**Comment by yinyangsx** [comment]
hi @bader, I disabled all tests except test_header, now sycl-cts post commit test is passed.

**Comment by yinyangsx** [comment]
hi @bader, do you know how to fix the failure of "OCL GEN9 LLVM Test Suite".

**Comment by yinyangsx** [comment]
hi @pvchupin, do you know how to fix the failure of LLVM test suite?

**Comment by bader** [comment]
> hi @bader, do you know how to fix the failure of "OCL GEN9 LLVM Test Suite".

@yinyangsx, this failure doesn't seem to be related to your change. Please, open a bug report for that.
I suppose that issue is sporadic, as I don't see it in other PR pre-commits, so merging with the head of the sycl branch might fix your pre-commit results.

**Comment by yinyangsx** [comment]
> > hi @bader, do you know how to fix the failure of "OCL GEN9 LLVM Test Suite".
> 
> @yinyangsx, this failure doesn't seem to be related to your change. Please, open a bug report for that. I suppose that issue is sporadic, as I don't see it in other PR pre-commits, so merging with the head of the sycl branch might fix your pre-commit results.

Hi @bader, I merged the head of sycl branch and run test but there still is a failed test case in "SYCL / Linux / OCL x64 LLVM Test Suite"
Failed Tests (1):
  SYCL :: DeviceLib/assert.cpp

**Comment by yanfeng3721** [comment]
Failed Tests (1):
  SYCL :: DeviceLib/assert.cpp([CMPLRTST-17873](https://jira.devtools.intel.com/browse/CMPLRTST-17873))

**Comment by yanfeng3721** [comment]
Failed Tests (1):
  SYCL :: DeviceLib/assert.cpp(https://jira.devtools.intel.com/browse/CMPLRTST-17874)
Flaky failure, rerun test pass.

**Comment by yanfeng3721** [comment]
@bader , @pvchupin , may I ask you merge the PR since all failure are analyzed? Thanks.

**Review by bader — CHANGES_REQUESTED** [review]
Looks good, but there are a few thing we can improve.

File: devops/actions/khronos_cts_test/action.yml [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+      sycl-ls --verbose
+      echo "::endgroup::"
+      echo "::group::SYCL_PI_TRACE=-1 sycl-ls"
+      echo $LD_LIBRARY_PATH
+      SYCL_PI_TRACE=-1 sycl-ls
+      echo "::endgroup::"
+      echo "::group::Run testing"
+      export SYCL_PI_CUDA_ENABLE_IMAGE_SUPPORT=1
+      export SYCL_DEVICE_FILTER=${{ matrix.sycl_device_filter }}
+      ./build/bin/test_all
```
Comment (bader) [inline devops/actions/khronos_cts_test/action.yml:115]:
```suggestion
      ./build/bin/test_all --list-devices
      ./build/bin/test_all
```

File: devops/cts_exclude_filter [hunk 3]
```diff
@@ -0,0 +1,52 @@
+accessor
+accessor_legacy
+address_space
+atomic
```
Comment (bader) [inline devops/cts_exclude_filter:4]: I hope you are going to enable all passing tests in follow-up PRs. Right?

Comment (yinyangsx) [reply devops/cts_exclude_filter:4]: Yes. of course.

File: devops/actions/khronos_cts_test/action.yml [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+      mkdir toolchain
+      tar -xf ${{ inputs.sycl_archive }} -C toolchain
+      rm -f ${{ inputs.sycl_archive }}
+  # TODO these must be in the container
+  - name: Build SYCL-CTS
+    shell: bash
+    run: |
+      echo "::group::CMake configuration"
+      export LD_LIBRARY_PATH=$PWD/toolchain/lib/:$LD_LIBRARY_PATH
+      export PATH=$PWD/toolchain/bin/:$PATH
```
[the remark this answers is not in the document]
Comment (bader) [reply devops/actions/khronos_cts_test/action.yml]: I figure out that it's needed to run sycl-ls.

**Review by bader — APPROVED** [review]

**Comment by bader** [comment]
Current status of issues for SYCL-CTS on NVIDIA GPU.

I started with building clean version of Khronos SYCL CTS:

Compiler commit - https://github.com/intel/llvm/pull/6439/commits/f55387001892d25c0017634134bcfce9efd1310a
Test log - https://github.com/intel/llvm/actions/runs/3072914412/jobs/4964801578
Following tests failed at compile step: context, device, event, exceptions, kernel, multi_ptr, platform, queue, reduction.
I think https://github.com/KhronosGroup/SYCL-CTS/pull/375 is going to address a part of these failures, but not all of them.

I've added all failing tests to the filter to exclude them from the build:
Compiler commit - https://github.com/intel/llvm/pull/6439/commits/c25703dbc104ccf6a6bcbb53e11f31c8918586d3
Test log - https://github.com/intel/llvm/actions/runs/3073539738/jobs/4965828019
optional_kernel_features failed at link step.

```
ptxas fatal   : Unresolved extern function '_Z18__spirv_AtomicIAddPyN5__spv5Scope4FlagENS0_19MemorySemanticsMask4FlagEy'
llvm-foreach: 
ptxas fatal   : Unresolved extern function '_Z18__spirv_AtomicIAddPyN5__spv5Scope4FlagENS0_19MemorySemanticsMask4FlagEy'
llvm-foreach: 
clang-16: error: ptxas command failed with exit code 255 (use -v to see invocation)
```

I've excluded optional_kernel_features from the build in:
Compiler commit - https://github.com/intel/llvm/pull/6439/commits/2d3e84d66b36c81d0752e390cc53ce899a4a97be
Test log - https://github.com/intel/llvm/actions/runs/3073804267/jobs/4966256628
Build failed with `ptxas fatal   : Internal error: reference to deleted symbol`. It was not clear which tests are causing this issue, so I started with disabling vector_* tests.

I've excluded vector tests from the build in:
Compiler commit - https://github.com/intel/llvm/pull/6439/commits/7d782f36ee2f09e0924fea45b9053697890c2096
Test log - https://github.com/intel/llvm/actions/runs/3074051359/jobs/4966656652
Unfortunately it didn't help and build failed with the same error message.

Next I excluded accessor and accessor_legacy tests from the build in:
Compiler commit - https://github.com/intel/llvm/pull/6439/commits/cb46c14142a59ab800fb035d1d7b1379fae5ac6d
Test log - https://github.com/intel/llvm/actions/runs/3074244346/jobs/4966997247
That fixed the build, but execution of kernel_bundle tests aborted.

```
/__w/llvm/llvm/khronos_sycl_cts/tests/kernel_bundle/../common/../../util/proxy.h:47: FAILED:
due to a fatal error condition:
  SIGABRT - Abort (abnormal termination) signal
```

I've excluded kernel_bundle tests from the build in:
Compiler commit - https://github.com/intel/llvm/pull/6439/commits/65160635848e2eeb31a76a7abe6b5d6f5849811b
Test log - https://github.com/intel/llvm/actions/runs/3074416382/jobs/4967284648
Tests were killed after 6 hours. The latest output was emitted by specialiazation_constants 

I've excluded specialization_constants tests from the build in:
Compiler commit - https://github.com/intel/llvm/pull/6439/commits/420f24bab30fe90168121e5b54b43623153790a0
Test log - https://github.com/intel/llvm/actions/runs/3075941019/jobs/4969835769
Tests pass and job took 23 minutes.

TODO:
1. Check if we can re-enable vector tests and one of accessor tests. AFAIK, these tests take a lot of time to compile and execute.
2. Enable CTS testing for other back-ends

File: devops/actions/khronos_cts_test/action.yml [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+  sycl_artifact:
+    description: 'Name of the artifact, that contains compiler toolchain'
+    required: true
+  sycl_archive:
+    description: 'Name of SYCL toolchain archive file'
+    required: false
+    default: 'llvm_sycl.tar.xz'
+  results_name_suffix:
+    description: 'Name suffix of the results artifact'
+    required: true
```
Comment (pvchupin) [inline devops/actions/khronos_cts_test/action.yml:17]: is it used?

File: devops/actions/khronos_cts_test/action.yml [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+      elif [ -e /opt/runtimes/oneapi-tbb/env/vars.sh ]; then
+        source /opt/runtimes/oneapi-tbb/env/vars.sh;
+      else
+        echo "no TBB vars in /opt/runtimes or /runtimes";
+      fi
+      # TODO remove workaround of FPGA emu bug
+      mkdir -p icd
+      echo /usr/lib/x86_64-linux-gnu/intel-opencl/libigdrcl.so > icd/gpu.icd
+      echo /runtimes/oclcpu/x64/libintelocl.so > icd/cpu.icd
+      echo /opt/runtimes/oclcpu/x64/libintelocl.so > icd/cpu2.icd
```
[the remark this answers is not in the document]
Comment (bader) [reply devops/actions/khronos_cts_test/action.yml:95]: It's needed for execution on OpenCL GPU/CPU devices. I can add these configuration right away. Are you okay with that?

File: devops/actions/khronos_cts_test/action.yml (hunk 5 above)
Comment (bader) [reply devops/actions/khronos_cts_test/action.yml]: I can't find any uses. Removed.

File: devops/actions/khronos_cts_test/action.yml [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+runs:
+  using: "composite"
+  steps:
+  - run: |
+      cp -r /actions .
+      git config --global --add safe.directory /__w/repo_cache/KhronosGroup/SYCL-CTS
+    shell: bash
+  - run: apt update && apt install -yqq opencl-headers ocl-icd-opencl-dev
+    shell: bash
+  - name: Checkout LLVM Test Suite
```
[the remark this answers is not in the document]
Comment (bader) [reply devops/actions/khronos_cts_test/action.yml]: Thanks, fixed.

File: devops/actions/khronos_cts_test/action.yml [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+    uses: actions/download-artifact@v2
+    with:
+      name: ${{ inputs.sycl_artifact }}
+  - name: Extract SYCL toolchain
+    shell: bash
+    run: |
+      mkdir toolchain
+      tar -xf ${{ inputs.sycl_archive }} -C toolchain
+      rm -f ${{ inputs.sycl_archive }}
+  # TODO these must be in the container
```
[the remark this answers is not in the document]
Comment (bader) [reply devops/actions/khronos_cts_test/action.yml]: I'm not sure. Let's remove this comment for now.

File: devops/scripts/generate_test_matrix.js [hunk 9]
```diff
@@ -69,6 +69,43 @@ module.exports = ({core, process}) => {
...
+      for (let [key, value] of Object.entries(inputs)) {
+        ctsString = ctsString.replaceAll("${{ inputs." + key + " }}", value);
+      }
+      if (needsDrivers) {
+        ctsString = ctsString.replaceAll(
+            "ghcr.io/intel/llvm/ubuntu2004_intel_drivers:latest",
+            "ghcr.io/intel/llvm/ubuntu2004_base:latest");
+      }
+
+      core.setOutput('cts_matrix', ctsString);
```
[the remark this answers is not in the document]
Comment (bader) [reply devops/scripts/generate_test_matrix.js:108]: I'm not sure if it make sense to "integrate" CTS code into llmv-test-suite. I see that @yinyangsx duplicated configuration. I guess it's done to be able to configure two independent execution environments for running llvm-test-suite and CTS.
I suggest we try to refactor this code in a separate PR. Are you okay with it?

File: devops/actions/khronos_cts_test/action.yml (hunk 6 above)
[the remark this answers is not in the document]
Comment (pvchupin) [reply devops/actions/khronos_cts_test/action.yml:91]: Let's start with CUDA first, but I think we want to add Level Zero GPU or OpenCL CPU, or maybe both as a reference.
Hard to estimate impact on the runners available.
Can be done in a separate patch.

File: devops/scripts/generate_test_matrix.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (pvchupin) [reply devops/scripts/generate_test_matrix.js:108]: I'm ok with separate PR. Thanks. I think it will be cleaner code after refactoring.

**Review by pvchupin — DISMISSED** [review]

File: devops/cts_exclude_filter [hunk 10]
```diff
@@ -0,0 +1,22 @@
+context
```
Comment (pvchupin) [inline devops/cts_exclude_filter:1]: This is CUDA only list at this point, right?
We may need a list per target?

Comment (bader) [reply devops/cts_exclude_filter:1]: Yes.
Yes.
